### PR TITLE
core: add ServerTransportFilter

### LINKED ISSUE #2132: Filter mechanism for server transport creation and deletion
To support a per-connection throttling mechanism used inside google, gRPC needs to allow user to add filters to the server that:
1. Gets called when a server transport is created and handshake-completed, and has access to transport-specific attributes such as client address and identity.
2. Adds per-transport attributes to ServerCall attributes.
3. Gets called when a server transport is terminated.

I discussed a few options with @ejona86:
#### ServerInterceptor

The filter returns a `ServerInterceptor` when transport is created. The interceptor is transport-scoped, unlike ordinary interceptors that are service-scoped.

``` java
interface ServerTransportFilter {
  ServerInterceptor transportCreated(SomePublicServerTransportType transport);
  void transportTerminated(SomePublicServerTransportType transport);
}
```

Pros:
1. Doesn't touch transport API (ServerTransport, ServerStream) which are internal.
2. Re-uses `ServerInterceptor` for purposes that `ServerInterceptor` is designed for (modifying ServerCall attributes etc).

Cons:
1. Changes the scope of `ServerInterceptor`. @ejona86 thinks bonding `ServerInterceptor` to a transport is seemingly generic but actually an ad-hoc solution for a particular problem. I don't necessarily agree though.
#### Decorating transport

The filter returns a decorated `ServerListener` at registration. Through a few more layers of wrapping, the filter would be able to decorate `ServerListener.transportCreated(),`ServerTransportListener.transportTerminated() and `ServerStream.attributes()`.

Pros:
1. Minimal API change

Cons:
1. Heavily coupled with transport API which is internal.
2. Duplicates the existing interceptor mechanism, but on `ServerStream` instead of `ServerCall`. It would be ideal to just decorate `ServerCall`, which is what `ServerInterceptor` does. `ServerInterceptor` currently is per-service, but our use case need it to be per-connection. 
3. Even more layers of decoration than the interceptors. The filter can be annoying to write.
#### Ad-hoc

Just for our original requirements, the filter would return per-transport attributes that are to be merged into ServerCall attributes.

``` java
interface ServerTransportFilter {
  Attributes transportCreated(SomePublicServerTransportType transport);
  void transportTerminated(SomePublicServerTransportType transport);
}
```

It probably won't work in this form, because the attributes such as peer security identity may not be available immediately after the transport is created, if the security handshake has not completed yet. On the other hand, the first two options can work around this situation by reading the attributes only when the first stream is created.

### PR DESCRIPTION
**Pull request by zhangkun83**
Called whenever a ServerTransport is ready and terminated. Has the ability to filter transport attributes, which `ServerCall.attributes()` are based on. Related changes:

- Attribute keys for remote address and SSL session are now moved from `ServerCall` to a neutral place `io.grpc.Grpc`, because they can also be used from `ServerTransportFilter`, and probably will be used on the client-side too.
- Added `transportReady()` to `ServerTransportListener`.

Resolves #2132 

/cc @lukaszx0

### CODE REVIEW

**Comment by zhangkun83** [comment]
@ejona86 Please review the PR as a whole.

File: core/src/main/java/io/grpc/CanonicalAttributes.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import io.grpc.Attributes;
+
+import java.net.SocketAddress;
+import javax.net.ssl.SSLSession;
+
+/**
+ * Canonical attribute keys that is used in various {@link Attributes} containers.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/1710")
+public final class CanonicalAttributes {
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/CanonicalAttributes.java]: Maybe add private constructor? I'm assuming the intention is to make it non initializable.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/CanonicalAttributes.java]: Done.

File: core/src/main/java/io/grpc/ServerCall.java [hunk 2]
```diff
@@ -57,21 +53,6 @@
 @DoNotMock("Use InProcessTransport and make a fake server instead")
 public abstract class ServerCall<ReqT, RespT> {
   /**
-   * {@link Attributes.Key} for the remote address of server call attributes
-   * {@link ServerCall#attributes()}
-   */
-  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1710")
-  public static final Attributes.Key<SocketAddress> REMOTE_ADDR_KEY =
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ServerCall.java]: Leave these in-place (just copying from `CanonicalAttributes`) with `@Deprecated` for one release cycle?

File: core/src/main/java/io/grpc/ServerTransportFilter.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Listens on server transport life-cycle events, with the capability to read and/or change
+ * transport attributes.  Attributes returned by this filter will be merged into {@link
+ * ServerCall#attributes}.
+ *
+ * <p>Multiple filters maybe registered to a server, in which case the output of a filter is the
+ * input of the next filter.  For example, what returned by {@link #transportReady} of a filter is
+ * passed to the same method of the next filter, and the last filter's return value is the effective
+ * transport attributes.  A filter should should the passed-in attributes instead of creating one
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ServerTransportFilter.java]: s/should should/should modify/

File: core/src/main/java/io/grpc/CanonicalAttributes.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import io.grpc.Attributes;
+
+import java.net.SocketAddress;
+import javax.net.ssl.SSLSession;
+
+/**
+ * Canonical attribute keys that are used in various {@link Attributes} containers.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/1710")
+public final class CanonicalAttributes {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/CanonicalAttributes.java]: Actually, this part seems orthogonal. We may want to split it out, so bikeshedding doesn't slow down the primary work.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/CanonicalAttributes.java]: I like 3. Basically I don't think these attributes are good fit for ClientCall/ServerCall, because they are also consumed by the filters, and the transport->filter data path doesn't have anything to do with ServerCall.

I can imagine `CallCredentials.ATTR_SECURITY_LEVEL` to be moved into `io.grpc.Grpc` if we will make it into ClientCall attributes.

Comment (zhangkun83) [inline core/src/main/java/io/grpc/CanonicalAttributes.java]: > Actually, this part seems orthogonal. We may want to split it out, so bikeshedding doesn't slow down the primary work.

Since this is part of the API, I would like to make the change early if we will do it eventually.

Comment (ejona86) [inline core/src/main/java/io/grpc/CanonicalAttributes.java]: I see. It isn't client vs server problem, but call vs transport. We can go with `io.grpc.Grpc` and see where it takes us.

File: core/src/main/java/io/grpc/internal/AbstractServerImplBuilder.java [hunk 5]
```diff
@@ -32,7 +32,9 @@
 package io.grpc.internal;
 
 import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkNotNull;
 
+import com.google.common.collect.ImmutableList;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/AbstractServerImplBuilder.java]: We're still avoiding guava collect for Android. I just got finished removing them. Could we remove the usage here as well, since it isn't too valuable?

File: core/src/main/java/io/grpc/ServerCall.java (hunk 2 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/ServerCall.java]: Done.

**Comment by zhangkun83** [comment]
@ejona86 Your comments are addressed. PTAL.

File: core/src/main/java/io/grpc/internal/ServerStream.java [hunk 6]
```diff
@@ -69,7 +69,8 @@
   void cancel(Status status);
 
   /**
-   * Attributes describing stream.
+   * Attributes describing stream.  This is inherited from the transport attributes, and used
+   * as the basis of {@link io.grpc.ServerCall#attributes}.
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerStream.java]: "as the basis" makes me think it may not be identical to `ServerCall.attributes()`. Are they not identical?

File: core/src/main/java/io/grpc/ServerCall.java [hunk 7]
```diff
@@ -59,15 +59,21 @@
   /**
    * {@link Attributes.Key} for the remote address of server call attributes
    * {@link ServerCall#attributes()}
+   *
+   * @deprecated use {@link io.grpc.Grpc#TRANSPORT_ATTR_REMOTE_ADDR} instead
    */
   @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1710")
+  @Deprecated
   public static final Attributes.Key<SocketAddress> REMOTE_ADDR_KEY =
           Attributes.Key.of("remote-addr");
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ServerCall.java]: This needs to be `Grpc.TRANSPORT_ATTR_REMOTE_ADDR`, otherwise it won't match. Similar below.

File: core/src/main/java/io/grpc/internal/ServerImpl.java [hunk 8]
```diff
@@ -126,6 +132,7 @@
     this.rootContext = Preconditions.checkNotNull(rootContext, "rootContext").fork();
     this.decompressorRegistry = decompressorRegistry;
     this.compressorRegistry = compressorRegistry;
+    this.transportFilters = Collections.unmodifiableList(transportFilters);
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: Make a copy as well.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java [hunk 9]
```diff
@@ -966,6 +972,8 @@ protected void assertX500SubjectDn(String tlsInfo) {
 
     assertEquals(1, certificates.size());
     assertEquals(tlsInfo, x509cert.getSubjectDN().toString());
+    assertEquals(serverCallCapture.get().attributes().get(Grpc.TRANSPORT_ATTR_SSL_SESSION),
```
Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: How is this check passing. Aren't the keys different instances right now?

Comment (ejona86) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: nm. I see. They are duped within the transport. Let's have one key a copy of the other, as I suggested above. That should be simpler.

**Comment by ejona86** [comment]
@zhangkun83 LGTM, but with some easy comments to address.

File: core/src/main/java/io/grpc/internal/ServerStream.java (hunk 6 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ServerStream.java]: Not necessarily identical. There may be per-stream attributes. Maybe.

File: core/src/main/java/io/grpc/internal/ServerImpl.java (hunk 8 above)
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: Done.

File: interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java (hunk 9 above)
Comment (zhangkun83) [inline interop-testing/src/main/java/io/grpc/testing/integration/AbstractInteropTest.java]: Yes that's much nicer.

**Comment by ejona86** [comment]
@zhangkun83 still LGTM

File: core/src/main/java/io/grpc/internal/ServerStream.java (hunk 6 above)
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerStream.java]: This is on the stream already, that's why it seemed a bit strange. No biggie.
